### DISCUSSION TITLE
Stacked Git 1.0 released

### DISCUSSION THREAD

**Discussion by jpgrayson** [stacked-git/stgit#96] [General]
At long last, StGit 1.0 has been released. :tada:

This release contains many important bug fixes, a few features, and a lot of foundational change under the hood.

## Stack Metadata

The stack metadata format is bumped to version 4. The huge change to stack metadata is that all of it is now contained in Git objects and *none of it* exists in regular files anymore. I.e. no more `.git/patches/`. Why does this matter? Two reasons:

1. StGit has had its stack metadata split across regular files (`.git/patches/`) and Git objects (e.g. `master.stgit`) for a long time. This has been a maintenance headache and a rich opportunity for bugs due to the large amount of parallel code required to maintain both metadata schemes.
2. Having the stack metadata entirely contained in the `<branch>.stgit` branch finally allows StGit stacks to be distributed using regular Git mechanism such as `git push` and `git pull`. It is early days for this capability--lots of room for improved tooling and new workflow ideas.

:warning: **Running StGit 1.0 will auto-upgrade your existing stack's metadata to version 4. Downgrading to an earlier version of StGit will require recreating stack and patches manually.**

## Python

StGit has dropped support for Python 2. This has also been a long time coming. While the upstream end-of-life of Python 2 made this decision somewhat easier, the main motivation was to reduce the considerable maintenance burden of maintaining Python 2+3 compatibility within the code. Dropping Python 2 allowed much compatibility code (some of which was pretty hacky) to be removed. The code was further updated to replace Python 2 idioms with contemporary Python 3 idioms.

The current minimum supported version is Python 3.5, which was first released in September, 2015 and has reached its end of life in September, 2020. Moving forward, StGit will seek to maintain compatibility with non-EOL versions of Python.

## Other

See the [release notes](https://github.com/stacked-git/stgit/releases/tag/v1.0) for a complete list of changes in StGit 1.0.

**Comment by ctmarinas** [comment]
Congratulations and many thanks Pete for getting StGit hit this milestone ;).

**Comment by NonLogicalDev** [comment]
@jpgrayson Thank you so much for your hard work!
